testsV2/postgres_test.go

```diff
@@ -56,5 +59,55 @@ func TestPostgresTcp(t *testing.T) {
 	requirePostgresVars(t)
 
 	dsn := fmt.Sprintf("user=%s password=%s database=%s sslmode=disable", *postgresUser, *postgresPass, *postgresDb)
-	proxyConnTest(t, *postgresConnName, "postgres", dsn, postgresPort, "")
+	proxyConnTest(t, context.Background(), []string{*postgresConnName}, "postgres", dsn, postgresPort, "")
+}
+
+func TestAuthWithToken(t *testing.T) {
+	if testing.Short() {
+		t.Skip("skipping Postgres integration tests")
+	}
+	requirePostgresVars(t)
+
+	ts, err := google.DefaultTokenSource(context.Background(), sqladmin.SqlserviceAdminScope)
+	if err != nil {
+		t.Errorf("failed to resolve token source: %v", err)
+	}
+	tok, err := ts.Token()
+	if err != nil {
+		t.Errorf("failed to get token: %v", err)
+	}
+
+	dsn := fmt.Sprintf("user=%s password=%s database=%s sslmode=disable",
+		*postgresUser, *postgresPass, *postgresDb)
+	proxyConnTest(t,
+		context.Background(),
+		[]string{"--token", tok.AccessToken, *postgresConnName},
+		"postgres", dsn, postgresPort, "")
+}
+
+func TestAuthWithCredentialsFile(t *testing.T) {
+	if testing.Short() {
+		t.Skip("skipping Postgres integration tests")
+	}
+	requirePostgresVars(t)
+
+	path, ok := os.LookupEnv("GOOGLE_APPLICATION_CREDENTIALS")
```

Comment:
Is it possible that maybe we can fake this one somehow? Like write a fake service account, set the env var just for this test, and check the header on our mocks? 

This is one integration test I'd prefer not to have given it means we have to have a service account key laying around.

Reply:
When we add a custom dialer, we could remove this. Until then, we don't have a way to inject a custom dialer, or HTTP client.

Reply:
So I was wrong: a custom dialer does not make this easy to test, as setting up a dialer also entails setting up authentication and so skips the code path where we append dialer options.

I also tried to remove the e2e test in favor of a more targeted unit test. That test was around the proxy.Client and involved setting up a mock HTTP server. This approach worked well for testing that the token is added to the header (I could just check the actual headers sent to the HTTP server).

For a test of the credentials file, however, the unit test works less well, as we need to also put together a fake OAuth2 server. When we pass a credential file dialer option, the underlying library reads the file and then tries to fetch a token at the URL specified in the credential file. While it is totally possible to assemble a fake OAuth2 server that returns tokens, I think the final resulting test depends on too many moving pieces, exercises a code path just below the top level, and ultimately doesn't provide as good test coverage.

So in the end, although I'd love to remove the service account dependency, a good and reliable test for this needs to be end to end. This is exactly the kind of feature that if it broke, it would have a very large impact on our customers, and so I want a robust test that guarantees the CLI flag, the parsing of args, the dialer options, and the ultimate authentication works. 

As a side note, if we switch to WIF, the [auth action](https://github.com/google-github-actions/auth) does support adding a `GOOGLE_APPLICATION_DEFAULT` credentials environment variable. So these tests would run in that setup as well.